samza-core/src/main/java/org/apache/samza/runtime/RemoteApplicationRunner.java

```diff
@@ -78,41 +89,47 @@ public void run(StreamApplication app) {
       plan.getJobConfigs().forEach(jobConfig -> {
           LOG.info("Starting job {} with config {}", jobConfig.getName(), jobConfig);
           JobRunner runner = new JobRunner(jobConfig);
+          jobRunners.add(runner);
           runner.run(true);
         });
+      jobRunnersByApp.put(app, jobRunners);
     } catch (Throwable t) {
       throw new SamzaException("Failed to run application", t);
     }
   }
 
   @Override
   public void kill(StreamApplication app) {
-    try {
-      ExecutionPlan plan = getExecutionPlan(app);
+    if (!jobRunnersByApp.containsKey(app)) {
```

Comment:
I don't think this will work. 

Take a look at run-app.sh and ApplicationRunnerMain.java. The current pattern is to invoke run-app.sh with an "operation" parameter to specify which command to run. Since this pattern starts a new JVM each time, this "jobRunnersByApp" will not survive between invocations. 

cc @nickpan47 and @xinyuiscool because I think they plan to change this in some way. 

But in the meantime, I think it's safer to replan the StreamApplication for each operation to get the JobRunner instances. Planning the same app should yield the same configs and therefore the same JobRunner behavior.

Reply:
Thanks for the review @jmakes 

I understand that it will not work if user uses run-app.sh to kill a job because it will be in a different/new process. But it seems like an existing issue, right? So I am not sure whether this patch will break anything.

This patch will improve will work if the remoteApplicationRunner.kill() is called in the same process after remoteApplicationRunner.run(). So it seems like an improvement. I can update the patch if there is better way to do it.

Not sure if I understand your last comment regarding replan". Maybe I can ask you in person about it.

Reply:
No, it's not an existing issue because before the patch, this kill method would always regenerate the ExecutionPlan. 

After the patch, it will fail because there is nothing in the map and no executionPlan from which we can recreate the JobRunner instances.

Reply:
@jmakes Thanks for the review. I found another way to trigger commit without having to explicitly stopping the application. I will close this PR and the JIRA ticket.